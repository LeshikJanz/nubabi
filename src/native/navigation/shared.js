import React from 'react';
import { pathOr } from 'ramda';
import LeftHeader from './LeftHeader';
import RightHeader from './RightHeader';

const babyNameTitle = ({ state }) => (
  pathOr('Nubabi', ['params', 'babyName'], state)
);

export function getBabyTitle() {
  return { title: babyNameTitle };
}

export const getTabHeaders = (navigate) => ({
  left: <LeftHeader navigate={navigate} />,
  right: <RightHeader navigate={navigate} />,
});
