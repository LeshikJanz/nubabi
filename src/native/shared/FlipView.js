import React, { Component } from 'react';
import {
  Animated,
  TouchableOpacity,
} from 'react-native';

const styles = {
  animatedContainer: {
    flex: 1,
  },
};

class FlipCard extends Component {
  constructor(props) {
    super(props);
    this._flipToggleCard = this._flipToggleCard.bind(this);
    this.state = {
      animatedValue: new Animated.Value(0),
      isFlipped: true,
    };
  }

  componentDidUpdate(prevProp, prevState) {
    if (this.state.isFlipped !== prevState.isFlipped) {
      this._flippedCard();
    }
  }

  _flipToggleCard() {
    this.setState({ isFlipped: !this.state.isFlipped });
  }

  _flippedCard() {
    Animated.spring(this.state.animatedValue, {
      toValue: 0,   // Returns to the start
      velocity: this.props.velocity,  // Velocity makes it move
      tension: this.props.tension, // Slow
      friction: this.props.friction,  // Oscillate a lot
    }).start();
  }

  flippedCardView(isFlipped) {
    if (isFlipped) {
      return this.props.renderFront;
    }
    return this.props.renderBack;
  }

  render() {
    const rotateX = this.state.animatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: ['0deg', '180deg'],
    });
    return (
      <TouchableOpacity onPress={this._flipToggleCard} style={styles.animatedContainer}>
        <Animated.View
          style={[styles.animatedContainer, { transform: [{ rotateX }] }]}
        >
          {this.flippedCardView(this.state.isFlipped)}
        </Animated.View>
      </TouchableOpacity>);
  }
}

FlipCard.propTypes = {
  renderBack: React.PropTypes.object.isRequired,
  renderFront: React.PropTypes.object.isRequired,
  velocity: React.PropTypes.number.isRequired,
  tension: React.PropTypes.number.isRequired,
  friction: React.PropTypes.number.isRequired,
};

module.exports = FlipCard;
