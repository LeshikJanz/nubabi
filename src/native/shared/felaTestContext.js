import React from 'react';
import { NativeFelaProvider } from '../components/FelaProvider';
import { createRenderer } from 'fela-native';
import theme from '../../common/themes/defaultTheme';

export const renderer = createRenderer();
const FelaProvider = NativeFelaProvider(renderer, theme);

const felaTestContext = component =>
  <FelaProvider>
    {component}
  </FelaProvider>;

export default felaTestContext;
