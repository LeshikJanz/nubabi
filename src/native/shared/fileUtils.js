// @flow
import { getContentTypeFromFilename } from '../../common/helpers/graphqlUtils';

export const isNewFile = (file: ?{ url: string }) => {
  if (!file) {
    return false;
  }

  return !file.url.startsWith('https://firebasestorage.googleapis.com');
};

// Parse results from react-native-image-picker into File interface inout
// (do not confuse with react-native-image-crop-picker which already gives
// us the expected result
export const fileFromImagePickerResult = file => {
  return {
    size: file.fileSize,
    name: file.fileName,
    url: file.uri,
    contentType: getContentTypeFromFilename(file.fileName),
  };
};
