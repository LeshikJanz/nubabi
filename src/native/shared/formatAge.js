import moment from 'moment';
import pluralize from 'pluralize';

const pluralizeAge = (count, unit) => {
  return [count, pluralize(unit, count), 'old'].join(' ');
};

export const formatAge = dobString => {
  const now = moment();
  const dob = moment(dobString);

  const diff = now.diff(dob, 'months');
  const shouldDisplayWeeks = diff === 0;

  if (shouldDisplayWeeks) {
    const weeks = now.diff(dob, 'week');
    return pluralizeAge(weeks, 'week');
  }

  const shouldDisplayYears = diff >= 12;

  if (shouldDisplayYears) {
    const years = now.diff(dob, 'years');
    return pluralizeAge(years, 'year');
  }

  return pluralizeAge(diff, 'month');
};
