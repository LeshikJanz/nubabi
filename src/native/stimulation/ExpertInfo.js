// @flow
import type { Expert, LayoutProps } from '../../common/types/index';
import React, { Component } from 'react';
import {
  View,
  Text,
  StyleSheet,
  Image,
  Dimensions,
  TouchableOpacity,
} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import { gql } from 'react-apollo';
import { path, head } from 'ramda';
import FlipCard from '../shared/FlipView';
import withLayout from '../components/withLayout';
import theme from '../../common/themes/defaultTheme';

type Props = {
  expert: Expert,
  activityDescription: string,
  layout: LayoutProps,
};

class ExpertInfo extends Component {
  props: Props;

  state = {
    isFlipped: false,
  };

  static fragments = {
    expert: gql`
      fragment ExpertInfo on Expert {
        id
        name
        discipline
        avatar {
          url
        }
        biography
      }
    `,
    activity: gql`
      fragment ExpertInfoActivity on Activity {
        introduction
      }
    `,
  };

  renderFront = onFlip => {
    const firstName = head(this.props.expert.name.split(' '));
    const avatar = { uri: path(['avatar', 'url'], this.props.expert) };

    return (
      <View style={styles.descriptionContainer}>
        <Text style={styles.descriptionHeader}>
          Our expert {firstName} says:
        </Text>
        <Image source={avatar} style={styles.avatar} />
        <Text style={styles.descriptionText}>
          {this.props.activityDescription}
        </Text>
        <View style={styles.expertDescriptionButton}>
          <Text style={styles.expertDescriptionName}>
            {this.props.expert.name}
          </Text>
          <Text style={styles.expertDescriptionProfession}>
            {this.props.expert.discipline}
          </Text>
          <TouchableOpacity style={styles.infoIcon} onPress={onFlip}>
            <Icon
              name="md-information-circle"
              size={24}
              color={theme.colors.primary}
            />
          </TouchableOpacity>
        </View>
      </View>
    );
  };

  renderBack = onFlip => {
    return (
      <View style={styles.biographyContainer}>
        <TouchableOpacity
          onPress={onFlip}
          style={{ alignSelf: 'flex-end', flex: 1 }}
        >
          <Icon name="ios-close-circle" color={theme.colors.gray} size={24} />
        </TouchableOpacity>
        <Image
          source={{ uri: this.props.expert.avatar.url }}
          style={styles.biographyAvatar}
        />
        <Text style={styles.expertBiographyName}>{this.props.expert.name}</Text>
        <Text style={styles.expertDescriptionProfession}>
          {this.props.expert.discipline}
        </Text>
        <Text style={styles.biographyText}>{this.props.expert.biography}</Text>
      </View>
    );
  };

  render() {
    const width = this.props.layout.viewportWidth ||
      Dimensions.get('window').width;

    return (
      <View
        style={{
          flex: 1,
          backgroundColor: theme.colors.open.white1,
          paddingTop: 20,
          marginTop: 0,
        }}
      >
        <FlipCard
          style={{ flex: 1 }}
          velocity={4}
          tension={7}
          friction={5}
          renderFront={onFlip => this.renderFront(onFlip)}
          renderBack={onFlip => this.renderBack(onFlip)}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  flipViewStyle: {
    flex: 1,
  },
  descriptionContainer: {
    marginLeft: 20,
    marginRight: 20,
    padding: 20,
    backgroundColor: '#fff',
    borderRadius: 4,
    shadowColor: '#000',
    shadowOpacity: 0.1,
    shadowRadius: 3,
    shadowOffset: {
      height: 2,
      width: 1,
    },
  },
  biographyContainer: {
    marginLeft: 20,
    marginRight: 20,
    padding: 20,
    backgroundColor: '#fff',
    borderRadius: 4,
    shadowColor: '#000',
    shadowOpacity: 0.1,
    shadowRadius: 3,
    shadowOffset: {
      height: 2,
      width: 1,
    },
    alignItems: 'center',
    flex: 1,
  },
  descriptionHeader: {
    color: '#454D56',
    fontSize: 18,
    marginBottom: 20,
  },
  avatar: {
    position: 'absolute',
    width: 40,
    height: 40,
    resizeMode: 'stretch',
    top: 10,
    right: 10,
    borderRadius: 20,
  },
  biographyAvatar: {
    width: 60,
    height: 60,
    resizeMode: 'stretch',
    borderRadius: 30,
  },
  descriptionText: {
    color: '#6A7683',
    fontSize: 16,
    lineHeight: 26,
    marginBottom: 10,
  },
  biographyText: {
    color: '#6A7683',
    fontSize: 14,
    lineHeight: 24,
    marginBottom: 10,
    marginTop: 20,
  },
  expertDescriptionButton: {
    paddingTop: 10,
    borderTopWidth: 1,
    borderColor: '#EDF1FA',
  },
  expertDescriptionName: {
    color: '#748294',
    fontSize: 14,
    fontWeight: '500',
  },
  expertBiographyName: {
    color: '#6A7683',
    fontSize: 14,
    fontWeight: '500',
    marginTop: 5,
  },
  expertDescriptionProfession: {
    color: '#748294',
    fontSize: 14,
    marginTop: 4,
  },
  infoIcon: {
    width: 20,
    height: 20,
    position: 'absolute',
    right: 10,
    top: 20,
    alignItems: 'flex-end',
    justifyContent: 'flex-end',
  },
  closeBiographyButton: {
    position: 'absolute',
    bottom: 20,
    right: 30,
    backgroundColor: 'transparent',
  },
});

export default withLayout(ExpertInfo);
