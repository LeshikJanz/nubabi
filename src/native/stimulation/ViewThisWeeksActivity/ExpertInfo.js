import React, { Component } from 'react';
import {
  View,
  Text,
  StyleSheet,
  Image,
  Dimensions,
} from 'react-native';
import _ from 'lodash';
import Icon from 'react-native-vector-icons/Ionicons';

import FlipCard from '../../shared/FlipView';
import { LIGHT_GREY } from '../../../common/themes/defaultTheme';

const width = Dimensions.get('window').width;
const infoIcon = require('../../../common/images/info_icon.png');

class ExpertInfo extends Component {
  constructor(props) {
    super(props);
    this.state = { isFlipped: false };
  }

  _renderFront = () => {
    const firstName = _.first(this.props.expert.name.split(' '));
    return (
      <View style={styles.descriptionContainer}>
        <Text style={styles.descriptionHeader}>Our expert {firstName} says:</Text>
        <Image source={this.props.expert.image_thumbnail} style={styles.avatar} />
        <Text style={styles.descriptionText}>{this.props.activity.description}</Text>
        <View style={styles.expertDescriptionButton}>
          <Text style={styles.expertDescriptionName}>{this.props.expert.name}</Text>
          <Text style={styles.expertDescriptionProfession}>{this.props.expert.profession}</Text>
          <Image style={styles.infoIcon} source={infoIcon} />
        </View>
      </View>
    );
  };

  _renderBack = () => {
    return (
      <View style={styles.biographyContainer}>
        <Image source={this.props.expert.image_thumbnail} style={styles.biographyAvatar} />
        <Text style={styles.expertBiographyName}>{this.props.expert.name}</Text>
        <Text style={styles.expertDescriptionProfession}>{this.props.expert.profession}</Text>
        <Text style={styles.biographyText}>{this.props.expert.biography}</Text>
        <Icon
          name="ios-close"
          size={30}
          color={LIGHT_GREY}
          style={styles.closeBiographyButton}
        />
      </View>
    );
  };

  render() {
    return (
      <FlipCard
        style={{ flex: 1 }}
        velocity={4}
        tension={7}
        friction={5}
        renderFront={this._renderFront()}
        renderBack={this._renderBack()}
      />
    );
  }
}

ExpertInfo.propTypes = {
  expert: React.PropTypes.object.isRequired,
  activity: React.PropTypes.object.isRequired,
};

const styles = StyleSheet.create({
  flipViewStyle: {
    flex: 1,
  },
  descriptionContainer: {
    marginLeft: 20,
    marginRight: 20,
    padding: 20,
    backgroundColor: '#fff',
    marginTop: 20,
    borderRadius: 4,
    shadowColor: '#000',
    shadowOpacity: 0.1,
    shadowRadius: 3,
    shadowOffset: {
      height: 2,
      width: 1,
    },

  },
  biographyContainer: {
    marginLeft: 20,
    marginRight: 20,
    marginTop: 20,
    padding: 20,
    backgroundColor: '#fff',
    borderRadius: 4,
    shadowColor: '#000',
    shadowOpacity: 0.1,
    shadowRadius: 3,
    shadowOffset: {
      height: 2,
      width: 1,
    },
    alignItems: 'center',
    flex: 1,
  },
  descriptionHeader: {
    color: '#454D56',
    fontSize: 18,
    marginBottom: 20,
  },
  avatar: {
    position: 'absolute',
    width: 40,
    height: 40,
    resizeMode: 'stretch',
    top: 10,
    right: 10,
    borderRadius: 20,
  },
  biographyAvatar: {
    width: 60,
    height: 60,
    resizeMode: 'stretch',
    borderRadius: 30,
  },
  descriptionText: {
    color: '#6A7683',
    fontSize: 16,
    lineHeight: 26,
    marginBottom: 10,
  },
  biographyText: {
    color: '#6A7683',
    fontSize: 14,
    lineHeight: 24,
    marginBottom: 10,
    marginTop: 20,
  },
  expertDescriptionButton: {
    paddingTop: 10,
    borderTopWidth: 1,
    borderColor: '#EDF1FA',
  },
  expertDescriptionName: {
    color: '#748294',
    fontSize: 14,
    fontWeight: '500',
  },
  expertBiographyName: {
    color: '#6A7683',
    fontSize: 14,
    fontWeight: '500',
    marginTop: 5,
  },
  expertDescriptionProfession: {
    color: '#748294',
    fontSize: 14,
    marginTop: 4,
  },
  infoIcon: {
    width: 20,
    height: 20,
    resizeMode: 'contain',
    position: 'absolute',
    right: 10,
    top: 20,
  },
  nextButtonsContainer: {
    marginTop: 10,
    backgroundColor: '#fff',
    width,
    height: 60,
    flexDirection: 'row',
  },
  closeBiographyButton: {
    position: 'absolute',
    bottom: 20,
    right: 30,
    backgroundColor: 'transparent',
  },
});

export default ExpertInfo;
