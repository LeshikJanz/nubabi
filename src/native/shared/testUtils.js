// @flow
import testRenderer from 'react-test-renderer';
import felaTestContext, { renderer as felaRenderer } from './felaTestContext';

const felaNativeToStyleObject = cache => {
  return Object.keys(cache).reduce((acc, key) => {
    return { ...acc, ...JSON.parse(key) };
  }, {});
};

export const createTestRenderer = () => (Component: any) => {
  const component = testRenderer.create(felaTestContext(Component));

  const snapshot = component.toJSON();
  snapshot.felaRules = felaNativeToStyleObject(felaRenderer.cache);

  return snapshot;
};

export const createExpectRender = () => {
  const render = createTestRenderer();
  return (Component: any) => {
    expect(render(Component)).toMatchSnapshot();
  };
};

export const expectRender = createExpectRender();
