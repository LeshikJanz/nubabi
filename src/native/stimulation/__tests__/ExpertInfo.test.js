import 'react-native';
import React from 'react';
import { ExpertInfo } from '../ExpertInfo';
import { expectRender } from '../../shared/testUtils';

test('it renders correctly', () => {
  const expert = {
    id: 1,
    name: 'Some Expert',
    discipline: 'Some Discipline',
    avatar: {
      url: 'http://example.com/example.png',
    },
    biography: 'some biography',
  };

  expectRender(
    <ExpertInfo
      expert={expert}
      activityDescription="Some activity description"
      layout={{ viewportWidth: 375 }}
    />,
  );
});
