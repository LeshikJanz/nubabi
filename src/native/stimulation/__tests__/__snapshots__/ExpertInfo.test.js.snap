// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
style = Object {}

<View
  style={
    Object {
      "backgroundColor": "#F8F9FC",
      "flex": 1,
      "marginTop": 0,
      "paddingTop": 20,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "transform": Array [
            Object {
              "rotateY": "0deg",
            },
            Object {
              "perspective": 2000,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#fff",
            "borderRadius": 4,
            "marginLeft": 20,
            "marginRight": 20,
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 1,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 3,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#454D56",
              "fontSize": 18,
              "marginBottom": 20,
            }
          }
        >
          Our expert 
          Some
           says:
        </Text>
        <Image
          resizeMode="stretch"
          source={
            Object {
              "uri": "http://example.com/example.png",
            }
          }
          style={
            Object {
              "borderRadius": 20,
              "height": 40,
              "position": "absolute",
              "right": 10,
              "top": 10,
              "width": 40,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#6A7683",
              "fontSize": 16,
              "lineHeight": 26,
              "marginBottom": 10,
            }
          }
        >
          Some activity description
        </Text>
        <View
          style={
            Object {
              "borderColor": "#EDF1FA",
              "borderTopWidth": 1,
              "paddingTop": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#748294",
                "fontSize": 14,
                "fontWeight": "500",
              }
            }
          >
            Some Expert
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#748294",
                "fontSize": 14,
                "marginTop": 4,
              }
            }
          >
            Some Discipline
          </Text>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "flex-end",
                "height": 20,
                "justifyContent": "flex-end",
                "opacity": 1,
                "position": "absolute",
                "right": 10,
                "top": 20,
                "width": 20,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#EA3154",
                    "fontSize": 24,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              ïŒ‘
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
